File: java/QuickSort.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;

/**
 * Implements full quick sort on a dataset and measures execution time
 * Uses last element as pivot as required
 */
public class QuickSort {
    private static class DataEntry implements Comparable<DataEntry> {
        int number;
        String text;
        
        DataEntry(int number, String text) {
            this.number = number;
            this.text = text;
        }
        
        @Override
        public int compareTo(DataEntry other) {
            return Integer.compare(this.number, other.number);
        }
        
        @Override
        public String toString() {
            return number + "," + text;
        }
    }

    /**
     * Reads all data entries from CSV file
     */
    private static List<DataEntry> readDataEntries(String filename) throws IOException {
        List<DataEntry> entries = new ArrayList<>();
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = reader.readLine()) != null) {
                String[] parts = line.split(",");
                if (parts.length == 2) {
                    entries.add(new DataEntry(Integer.parseInt(parts[0]), parts[1]));
                }
            }
        }
        return entries;
    }

    /**
     * Writes sorted entries to CSV file
     */
    private static void writeSortedData(List<DataEntry> entries, String filename) throws IOException {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
            for (DataEntry entry : entries) {
                writer.write(entry.toString() + "\n");
            }
        }
    }

    /**
     * Quick sort implementation
     */
    private static void quickSort(List<DataEntry> entries) {
        quickSort(entries, 0, entries.size() - 1);
    }

    private static void quickSort(List<DataEntry> entries, int low, int high) {
        if (low < high) {
            int partitionIndex = partition(entries, low, high);
            
            // Recursively sort elements before and after partition
            quickSort(entries, low, partitionIndex - 1);
            quickSort(entries, partitionIndex + 1, high);
        }
    }

    private static int partition(List<DataEntry> entries, int low, int high) {
        // pivot (Element to be placed at right position)
        DataEntry pivot = entries.get(high);
        
        int i = low - 1; // Index of smaller element
        
        for (int j = low; j < high; j++) {
            // If current element is smaller than or equal to pivot
            if (entries.get(j).compareTo(pivot) <= 0) {
                i++;
                
                // swap arr[i] and arr[j]
                DataEntry temp = entries.get(i);
                entries.set(i, entries.get(j));
                entries.set(j, temp);
            }
        }
        
        // swap arr[i+1] and arr[high] (or pivot)
        DataEntry temp = entries.get(i + 1);
        entries.set(i + 1, entries.get(high));
        entries.set(high, temp);
        
        return i + 1;
    }

    public static void main(String[] args) {
        if (args.length != 1) {
            System.out.println("Usage: java QuickSort <input_file>");
            return;
        }
        
        try {
            String inputFile = args[0];
            List<DataEntry> entries = readDataEntries(inputFile);
            
            // Measure execution time (excluding I/O)
            long startTime = System.nanoTime();
            quickSort(entries);
            long endTime = System.nanoTime();
            double duration = (endTime - startTime) / 1_000_000.0; // milliseconds
            
            // Write sorted data
            String outputFile = "quick_sort_" + entries.size() + ".csv";
            writeSortedData(entries, outputFile);
            
            System.out.println("Quick sort completed in " + duration + " ms");
            System.out.println("Sorted data saved to: " + outputFile);
        } catch (IOException e) {
            System.err.println("Error processing files: " + e.getMessage());
        }
    }
}

File: python/quick_sort.py
import os
import sys
import time

# Configuration
CHUNK_SIZE = 5_000_000  # Safe for 16GB RAM, improves speed
TEMP_DIR = "temp_chunks"
OUTPUTS_FOLDER = "../outputs"
DATA_FOLDER = "../datasets"

class Record:
    def __init__(self, num, line):
        self.num = num
        self.line = line

    def __lt__(self, other):
        return self.num < other.num

# Step 1: Split the input file into smaller sorted chunks
def split_and_sort(input_file_path):
    os.makedirs(TEMP_DIR, exist_ok=True)
    chunk_files = []
    records = []
    index = 0

    with open(input_file_path, 'r') as src:
        for ln_num, ln in enumerate(src, 1):
            parts = ln.strip().split(',', 1)
            if len(parts) == 2:
                try:
                    num = int(parts[0])
                    rest = parts[1]
                    records.append(Record(num, rest))
                except ValueError:
                    continue

            if len(records) >= CHUNK_SIZE:
                dump_chunk(records, index, chunk_files)
                records.clear()
                index += 1

        if records:
            dump_chunk(records, index, chunk_files)

    return chunk_files

# Sort and save each chunk to disk using iterative quicksort
def dump_chunk(items, idx, out_files):
    quicksort_iterative(items)
    path = os.path.join(TEMP_DIR, f"chunk_{idx}.csv")
    with open(path, 'w', buffering=1_048_576) as f:
        for entry in items:
            f.write(f"{entry.num},{entry.line}\n")
    out_files.append(path)

# Iterative quicksort avoids recursion limit issues
def quicksort_iterative(arr):
    stack = [(0, len(arr) - 1)]
    while stack:
        lo, hi = stack.pop()
        if lo < hi:
            pivot_idx = partition(arr, lo, hi)
            stack.append((lo, pivot_idx - 1))
            stack.append((pivot_idx + 1, hi))

# Partitioning using last element as pivot 
def partition(arr, low, high):
    pivot_val = arr[high].num
    i = low - 1
    for j in range(low, high):
        if arr[j].num < pivot_val:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

# Merge sorted chunks manually without using heapq
def combine_chunks(sorted_files, final_output_path):
    open_files = [open(path, 'r') for path in sorted_files]
    buffers = []

    for f in open_files:
        line = f.readline()
        if line:
            parts = line.strip().split(',', 1)
            if len(parts) == 2:
                buffers.append((int(parts[0]), parts[1], f))
            else:
                buffers.append((float('inf'), '', f))
        else:
            buffers.append((float('inf'), '', f))

    with open(final_output_path, 'w', buffering=1_048_576) as out:
        while True:
            min_index = -1
            min_val = float('inf')

            for i, (val, _, _) in enumerate(buffers):
                if val < min_val:
                    min_val = val
                    min_index = i

            if min_index == -1:
                break  # All files exhausted

            num, txt, f = buffers[min_index]
            out.write(f"{num},{txt}\n")

            nxt = f.readline()
            if nxt:
                parts = nxt.strip().split(',', 1)
                if len(parts) == 2:
                    buffers[min_index] = (int(parts[0]), parts[1], f)
                else:
                    buffers[min_index] = (float('inf'), '', f)
            else:
                buffers[min_index] = (float('inf'), '', f)

    for f in open_files:
        f.close()
    for f in sorted_files:
        os.remove(f)
    os.rmdir(TEMP_DIR)

# Generate output file path from input filename
def make_output_path(fname):
    os.makedirs(OUTPUTS_FOLDER, exist_ok=True)
    label = fname.replace("dataset_", "").replace(".csv", "")
    return os.path.join(OUTPUTS_FOLDER, f"quick_sort_{label}.csv")

# Main entry point of the script
def main():
    if len(sys.argv) < 2:
        sys.exit(1)

    file_name = sys.argv[1]
    full_path = os.path.join(DATA_FOLDER, file_name)

    if not os.path.exists(full_path):
        sys.exit(1)

    start = time.time()
    pieces = split_and_sort(full_path)
    output_file = make_output_path(file_name)
    combine_chunks(pieces, output_file)
    end = time.time()

    print(int((end - start) * 1000))  # Total sort runtime in milliseconds

if __name__ == "__main__":
    main()
